pageserver/src/http/routes.rs

```diff
@@ -495,13 +495,39 @@ async fn get_lsn_by_timestamp_handler(
     let timeline = active_timeline_of_active_tenant(tenant_id, timeline_id).await?;
     let result = timeline.find_lsn_for_timestamp(timestamp_pg, &ctx).await?;
 
-    let result = match result {
-        LsnForTimestamp::Present(lsn) => format!("{lsn}"),
-        LsnForTimestamp::Future(_lsn) => "future".into(),
-        LsnForTimestamp::Past(_lsn) => "past".into(),
-        LsnForTimestamp::NoData(_lsn) => "nodata".into(),
-    };
-    json_response(StatusCode::OK, result)
+    if request
+        .headers()
+        .get(header::ACCEPT)
+        .map(|v| v == "application/json")
+        .unwrap_or_default()
+    {
+        #[derive(serde::Serialize)]
+        struct Result {
+            lsn: String,
+            kind: &'static str,
```

Comment:
serde could probably get this off an `enum Result`, the variant name. Then the LsnForTimestamp could be serialized directly.

I don't have a suggestion for this yet, can show example later.

Reply:
I don't really agree that we should use serde here as it makes the code harder to understand: instead of looking at one place that contains the serialization code, you now have to look all around the code base to understand what is being serialized.

I would understand it if the user of this API were also Rust, and we had a common `api` crate or such which contains all the types of the REST API, so we don't care much about the actual json outside of backward compatibility concerns. That's the area serde is strong in, just add a derive and forget about the wire format. For creating json formatted in a specific way, one needs a lot of special knowledge of all the tricks that serde's complex machinery for serialization and deserialization contains.

But here, the format is specified in an yml format and the user of the API is either the python tests, or components written in go.

Reply:
I had a look and you cannot use `serde_with` with enums, so we cannot do this at this point.